File: app/src/main/java/ca/fuwafuwa/kaku/Windows/Views/KanjiCharacterView.java
package ca.fuwafuwa.kaku.Windows.Views;

import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.util.Log;
import android.util.TypedValue;
import android.view.GestureDetector;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;

import ca.fuwafuwa.kaku.KakuTools;
import ca.fuwafuwa.kaku.Ocr.OcrChar;
import ca.fuwafuwa.kaku.R;

/**
 * Created by 0xbad1d3a5 on 5/5/2016.
 */
public class KanjiCharacterView extends TextView implements GestureDetector.OnGestureListener {

    private static final String TAG = KanjiCharacterView.class.getName();

    private Context mContext;
    private Point mOrigPos;
    private SquareGridView.SquareViewListener mCallback;
    private GestureDetector mGestureDetector;
    private MotionEvent mScrollStartEvent;
    private OcrChar mOcrChar;

    private int mSizePx;
    private int mTextSizeDp;
    private int mCharPos;
    private int[] mOrigPosRaw;

    private boolean mResetOrigPos;
    private boolean mEdited;

    public KanjiCharacterView(Context context) {
        super(context);
        Init(context);
    }

    public KanjiCharacterView(Context context, AttributeSet attrs) {
        super(context, attrs);
        Init(context);
    }

    public KanjiCharacterView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        Init(context);
    }

    public KanjiCharacterView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        Init(context);
    }

    private void Init(Context context){
        mContext = context;
        mGestureDetector = new GestureDetector(mContext, this);
        mOrigPos = new Point();
        mOrigPosRaw = new int[2];
        mResetOrigPos = true;
        mSizePx = KakuTools.dpToPx(mContext, 35);
        mTextSizeDp = 20;
        mEdited = false;
    }

    public void setKanjiViewCallback(SquareGridView.SquareViewListener callback){
        mCallback = callback;
    }

    public void recycle()
    {
        mCallback = null;
        mGestureDetector = null;
    }

    public void setOcrChar(OcrChar ocrChar){
        mOcrChar = ocrChar;
    }

    public OcrChar getOcrChar(){
        return mOcrChar;
    }

    public int[] getOrigPosRaw(){
        return mOrigPosRaw;
    }

    public int getCharPos() {
        return mCharPos;
    }

    public void setCharPos(int charPos) {
        mCharPos = charPos;
    }

    public void setSize(int dp){
        mSizePx = KakuTools.dpToPx(mContext, dp);
    }

    public void setTextSize(int dp){
        mTextSizeDp = dp;
    }

    public boolean isEdited() {
        return mEdited;
    }

    public void setEdited(boolean edited) {
        mEdited = edited;
    }

    public void removeBackground(){
        setBackground(null);
    }

    public void setBackground(int id){
        Drawable bg = mContext.getDrawable(id);
        setBackground(bg);
    }

    public void setBackground(){
        Drawable bg = mContext.getDrawable(R.drawable.bg_translucent_border_0_blue_blue);
        setBackground(bg);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec){
        setTextSize(TypedValue.COMPLEX_UNIT_DIP, mTextSizeDp);
        setTextColor(Color.BLACK);
        setBackgroundColor(Color.WHITE);
        setGravity(Gravity.CENTER);
        setMeasuredDimension(mSizePx, mSizePx);
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);

        if (mResetOrigPos){
            getLocationOnScreen(mOrigPosRaw);
            mOrigPos.x = (int) getX();
            mOrigPos.y = (int) getY();
            mResetOrigPos = false;
        }
    }

    @Override
    protected void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        mResetOrigPos = true;
    }

    @Override
    public boolean onTouchEvent(MotionEvent e){

        mGestureDetector.onTouchEvent(e);

        if (e.getAction() == MotionEvent.ACTION_UP){
            setX(mOrigPos.x);
            setY(mOrigPos.y);
            setVisibility(View.VISIBLE);
            mCallback.onSquareScrollEnd(this, e);
            mScrollStartEvent = null;
        }

        return true;
    }

    @Override
    public boolean onDown(MotionEvent motionEvent) {
        Log.d(TAG, String.format("%s: onDown: %f x %f", getText(), getX(), getY()));
        return true;
    }

    @Override
    public void onShowPress(MotionEvent motionEvent) {

    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        Log.d(TAG, "onSingleTapUp");
        mCallback.onSquareTouch(this);
        return true;
    }

    @Override
    public boolean onScroll(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {

        if (mScrollStartEvent == null){
            mScrollStartEvent = motionEvent;
            mCallback.onSquareScrollStart(this, mScrollStartEvent);
        }
        else {
            setX(motionEvent1.getRawX() - motionEvent.getRawX() + mOrigPos.x);
            setY(motionEvent1.getRawY() - motionEvent.getRawY() + mOrigPos.y);
            setVisibility(View.INVISIBLE);
            mCallback.onSquareScroll(this, motionEvent, motionEvent1);
        }

        return true;
    }

    @Override
    public void onLongPress(MotionEvent motionEvent) {
        Log.d(TAG, "onLongPress");
    }

    @Override
    public boolean onFling(MotionEvent motionEvent, MotionEvent motionEvent1, float v, float v1) {
        Log.d(TAG, "onFling");
        return false;
    }
}

File: app/src/main/java/ca/fuwafuwa/kaku/Windows/Views/KanjiCharacterView.kt
package ca.fuwafuwa.kaku.Windows.Views

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.util.Log
import android.util.TypedValue
import android.view.GestureDetector
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import ca.fuwafuwa.kaku.*
import ca.fuwafuwa.kaku.Ocr.BoxParams
import ca.fuwafuwa.kaku.Windows.*
import ca.fuwafuwa.kaku.Windows.Data.DisplayDataOcr
import ca.fuwafuwa.kaku.Windows.Data.ISquareChar
import ca.fuwafuwa.kaku.Windows.Interfaces.ICopyText
import ca.fuwafuwa.kaku.Windows.Interfaces.IRecalculateKanjiViews
import ca.fuwafuwa.kaku.Windows.Interfaces.ISearchPerformer

/**
 * Created by 0xbad1d3a5 on 5/5/2016.
 */
class KanjiCharacterView : FrameLayout, GestureDetector.OnGestureListener, IRecalculateKanjiViews
{
    private lateinit var mContext: Context
    private lateinit var mGestureDetector: GestureDetector
    private lateinit var mWindowCoordinator: WindowCoordinator
    private lateinit var mSearchPerformer: ISearchPerformer
    private lateinit var mKanjiChoiceWindow: KanjiChoiceWindow
    private lateinit var mEditWindow: EditWindow
    private lateinit var mSquareChar: ISquareChar

    private lateinit var mKanjiTextView: TextView
    private lateinit var mIconImageView: ImageView

    private var mCellSizePx: Int = 0
    private var mScrollStartEvent: MotionEvent? = null

    constructor(context: Context) : super(context)
    {
        Init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)
    {
        Init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)
    {
        Init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)
    {
        Init(context)
    }

    private fun Init(context: Context)
    {
        mContext = context
        mGestureDetector = GestureDetector(mContext, this)

        mKanjiTextView = TextView(mContext)
        mKanjiTextView.gravity = Gravity.CENTER
        mKanjiTextView.setTextSize(TypedValue.COMPLEX_UNIT_DIP, 20.toFloat())
        mKanjiTextView.setTextColor(Color.BLACK)
        addView(mKanjiTextView)

        mIconImageView = ImageView(mContext)
        mIconImageView.visibility = INVISIBLE
        addView(mIconImageView)
    }

    fun getSquareChar(): ISquareChar
    {
        return mSquareChar
    }

    fun setDependencies(windowCoordinator: WindowCoordinator, searchPerformer: ISearchPerformer)
    {
        mWindowCoordinator = windowCoordinator
        mSearchPerformer = searchPerformer

        mKanjiChoiceWindow = mWindowCoordinator.getWindowOfType(WINDOW_KANJI_CHOICE)
        mEditWindow = mWindowCoordinator.getWindowOfType(WINDOW_EDIT)
    }

    fun setText(squareChar: ISquareChar)
    {
        mSquareChar = squareChar
        mKanjiTextView.text = squareChar.char
    }

    fun setCellSize(px: Int)
    {
        mCellSizePx = dpToPx(context, pxToDp(context, px) - 2)
    }

    fun highlight()
    {
        background = mContext.getDrawable(R.drawable.bg_translucent_border_0_blue_blue)
    }

    fun highlightLight()
    {
        background = mContext.getDrawable(R.drawable.bg_transparent_border_0_nil_default)
    }

    fun unhighlight()
    {
        background = null
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int)
    {
        val cellWidthSpec = View.MeasureSpec.makeMeasureSpec(mCellSizePx, View.MeasureSpec.EXACTLY)
        val cellHeightSpec = View.MeasureSpec.makeMeasureSpec(mCellSizePx, View.MeasureSpec.EXACTLY)

        for (i in 0 until childCount)
        {
            getChildAt(i).measure(cellWidthSpec, cellHeightSpec)
        }

        setMeasuredDimension(mCellSizePx, mCellSizePx)
    }

    override fun onTouchEvent(e: MotionEvent): Boolean
    {
        mGestureDetector.onTouchEvent(e)

        if (e.action == MotionEvent.ACTION_UP)
        {
            visibility = View.VISIBLE

            if (mScrollStartEvent != null)
            {
                mScrollStartEvent = null

                mKanjiTextView.visibility = View.VISIBLE
                mIconImageView.visibility = View.INVISIBLE

                val choiceResult = mKanjiChoiceWindow.onSquareScrollEnd(e)
                when (choiceResult.first)
                {
                    ChoiceResultType.SWAP ->
                    {
                        mKanjiTextView.text = choiceResult.second
                        mSquareChar.text = choiceResult.second
                        recalculateKanjiViews()
                    }
                    ChoiceResultType.EDIT ->
                    {
                        val window = getProperWindow<Window>()
                        if (mSquareChar.displayData is DisplayDataOcr)
                        {
                            window.hide()
                        }

                        mEditWindow.setInfo(mSquareChar)
                        mEditWindow.setInputDoneCallback(this)
                        mEditWindow.show()
                    }
                    ChoiceResultType.DELETE ->
                    {
                        mSquareChar.text = ""
                        recalculateKanjiViews()
                    }
                    ChoiceResultType.NONE ->
                    {
                        // Do nothing
                    }
                }
            }
        }

        return true
    }

    override fun recalculateKanjiViews()
    {
        val cwindow = getProperWindow<IRecalculateKanjiViews>()
        cwindow.recalculateKanjiViews()

        val window = getProperWindow<Window>()
        window.show()
    }

    override fun onSingleTapUp(e: MotionEvent): Boolean
    {
        highlightLight()
        mSquareChar.userTouched = true
        mSearchPerformer.performSearch(mSquareChar)
        return true
    }

    override fun onScroll(motionEvent: MotionEvent, motionEvent1: MotionEvent, v: Float, v1: Float): Boolean
    {
        // scroll event start
        if (mScrollStartEvent == null)
        {
            Log.d(TAG, "ScrollStart")
            mScrollStartEvent = motionEvent

            unhighlight()
            mKanjiTextView.visibility = View.INVISIBLE
            mIconImageView.visibility = View.VISIBLE
            mIconImageView.setImageResource(R.drawable.icon_swap)

            mKanjiChoiceWindow.onSquareScrollStart(mSquareChar, getKanjiBoxParams())
        }
        // scroll event continuing
        else {
            Log.d(TAG, "ScrollContinue")
            mIconImageView.setImageResource(mKanjiChoiceWindow.onSquareScroll(motionEvent1))
        }

        return true
    }

    override fun onDown(motionEvent: MotionEvent): Boolean
    {
        return false
    }

    override fun onFling(motionEvent: MotionEvent, motionEvent1: MotionEvent, v: Float, v1: Float): Boolean
    {
        return false
    }

    override fun onLongPress(motionEvent: MotionEvent)
    {
        val window = getProperWindow<ICopyText>()
        window.copyText()
    }

    override fun onShowPress(e: MotionEvent?)
    {
    }

    private fun <WindowType> getProperWindow() : WindowType
    {
        return if (mSquareChar.displayData.instantMode)
        {
            mWindowCoordinator.getWindowOfType(WINDOW_INSTANT_KANJI)
        }
        else {
            mWindowCoordinator.getWindowOfType(WINDOW_INFO)
        }
    }

    private fun getKanjiBoxParams() : BoxParams
    {
        var pos = IntArray(2)
        getLocationOnScreen(pos)
        return BoxParams(pos[0], pos[1], width, height)
    }

    companion object
    {
        private val TAG = KanjiCharacterView::class.java.name
    }
}
